File: server/Tests/TestSMTP.swift
import XCTest
@testable import Joshinkan

final class UserTests: XCTestCase {
  func testParseDescription_missingClosingCaret() {
    XCTAssertThrowsError(try User(description: "<missing")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.UNEXPECTED_FORMAT)
    }
  }
  
  func testParseDescription_missingOpeningCaret() {
    XCTAssertThrowsError(try User(description: "missing>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.UNEXPECTED_FORMAT)
    }
  }
  
  func testParseDescription_textAfterEmail() {
    XCTAssertThrowsError(try User(description: "John Smith <john@example.com> shouldnotbehere")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.UNEXPECTED_FORMAT)
    }
  }
  
  func testParseDescription_caretsInName() {
    XCTAssertThrowsError(try User(description: "Jo<hn Smith <john@example.com>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
    
    XCTAssertThrowsError(try User(description: "Jo>hn Smith <john@example.com>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.UNEXPECTED_FORMAT)
    }
    
    XCTAssertThrowsError(try User(description: "J<o>hn Smith <john@example.com>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
  }
  
  func testParseDescription_invalidEmail() {
    XCTAssertThrowsError(try User(description: "John Smith")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
    
    XCTAssertThrowsError(try User(description: "invalid@invalid")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
    
    XCTAssertThrowsError(try User(description: "<John Smith>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
    
    XCTAssertThrowsError(try User(description: "<invalid@i.i>")) { error in
      XCTAssertEqual(error as! User.ParsingError, User.ParsingError.INVALID_EMAIL)
    }
  }

  
  func testParseDescription_fullDescription() {
    let user = try! User(description: "John Smith <john@example.com>")
    XCTAssertEqual(user.email, "john@example.com")
    XCTAssertEqual(user.name, "John Smith")
  }
  
  func testParseDescription_fullDescriptionExtraWhitespace() {
    let user = try! User(description: "John Smith    <john@example.com>      ")
    XCTAssertEqual(user.email, "john@example.com")
    XCTAssertEqual(user.name, "John Smith")
  }
  
  func testParseDescription_emailOnly() {
    let user = try! User(description: "<john@example.com>")
    XCTAssertEqual(user.email, "john@example.com")
    XCTAssertNil(user.name)
  }
  
  func testParseDescription_emailAsName() {
    let user = try! User(description: "john@example.com")
    XCTAssertEqual(user.email, "john@example.com")
    XCTAssertNil(user.name)
  }
}

File: server/test/test_smtp.py
import pytest
from joshinkan.smtp import EmailUser


def test_parse_description_missing_closing_caret():
    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.UNEXPECTED_FORMAT)
    ):
        EmailUser.from_description("<missing")


def test_parse_description_missing_opening_caret():
    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.UNEXPECTED_FORMAT)
    ):
        EmailUser.from_description("missing>")


def test_parse_description_text_after_email():
    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.UNEXPECTED_FORMAT)
    ):
        EmailUser.from_description("John Smith <john@example.com> shouldnotbehere")


def test_parse_description_carets_in_name():
    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("Jo<hn Smith <john@example.com>")

    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.UNEXPECTED_FORMAT)
    ):
        EmailUser.from_description("Jo>hn Smith <john@example.com>")

    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("J<o>hn Smith <john@example.com>")


def test_parse_description_invalid_email():
    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("John Smith")

    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("invalid@invalid")

    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("<John Smith>")

    with pytest.raises(
        EmailUser.ParsingError, match=str(EmailUser.ParsingError.INVALID_EMAIL)
    ):
        EmailUser.from_description("<invalid@i.i>")


def test_parse_description_full_description():
    user = EmailUser.from_description("John Smith <john@example.com>")
    assert user.email == "john@example.com"
    assert user.name == "John Smith"


def test_parse_description_full_description_extra_whitespace():
    user = EmailUser.from_description("John Smith    <john@example.com>      ")
    assert user.email == "john@example.com"
    assert user.name == "John Smith"


def test_parse_description_email_only():
    user = EmailUser.from_description("<john@example.com>")
    assert user.email == "john@example.com"
    assert user.name is None


def test_parse_description_email_as_name():
    user = EmailUser.from_description("john@example.com")
    assert user.email == "john@example.com"
    assert user.name is None
